htlcswitch/circuit_map.go

```diff
@@ -0,0 +1,848 @@
+			circuitBytes := bs[i].Bytes()
+
+			err := circuitBkt.Put(inKeyBytes, circuitBytes)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+
+	// Return if the write succeeded.
+	if err == nil {
+		actions.Adds = adds
+		actions.Drops = drops
+		actions.Fails = fails
+		return actions, nil
+	}
+
+	// Otherwise, rollback the circuits added to the pending set if the
+	// write failed.
+	cm.mtx.Lock()
+	for _, circuit := range adds {
+		inKey := circuit.InKey()
+		if _, ok := cm.pending[inKey]; !ok {
+			delete(cm.pending, inKey)
+		}
+	}
+	cm.mtx.Unlock()
+
+	actions.Drops = drops
+	actions.Fails = addFails
+
+	return actions, err
+}
+
+// Keystone is a tuple binding an incoming and outgoing CircuitKey. Keystones
+// are preemptively written by an outgoing link before signing a new commitment
+// state, and cements which HTLCs we are awaiting a response from a remote peer.
+type Keystone struct {
+	InKey  CircuitKey
+	OutKey CircuitKey
+}
+
+// String returns a human readable description of the Keystone.
+func (k *Keystone) String() string {
+	return fmt.Sprintf("%s --> %s", k.InKey, k.OutKey)
+}
+
+// OpenCircuits sets the outgoing circuit key for the circuit identified by
+// inKey, persistently marking the circuit as opened. After the changes have
+// been persisted, the circuit map's in-memory indexes are updated so that this
+// circuit can be queried using LookupByKeystone or LookupByPaymentHash.
+func (cm *circuitMap) OpenCircuits(keystones ...Keystone) error {
+	if len(keystones) == 0 {
+		return nil
+	}
+
+	// Check that all keystones correspond to committed-but-unopened
+	// circuits.
+	cm.mtx.RLock()
+	openedCircuits := make([]*PaymentCircuit, 0, len(keystones))
+	for _, ks := range keystones {
+		if _, ok := cm.opened[ks.OutKey]; ok {
+			cm.mtx.RUnlock()
+			return ErrDuplicateKeystone
+		}
+
+		circuit, ok := cm.pending[ks.InKey]
+		if !ok {
+			cm.mtx.RUnlock()
+			return ErrUnknownCircuit
+		}
+
+		openedCircuits = append(openedCircuits, circuit)
+	}
+	cm.mtx.RUnlock()
+
+	err := cm.db.Update(func(tx *bolt.Tx) error {
+		// Now, load the circuit bucket to which we will write the
+		// already serialized circuit.
+		keystoneBkt := tx.Bucket(circuitKeystoneKey)
+		if keystoneBkt == nil {
+			return ErrCorruptedCircuitMap
+		}
+
+		for _, ks := range keystones {
+			outBytes := ks.OutKey.Bytes()
+			inBytes := ks.InKey.Bytes()
+			err := keystoneBkt.Put(outBytes, inBytes)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	cm.mtx.Lock()
+	for i, circuit := range openedCircuits {
+		ks := keystones[i]
+
+		// Since our persistent operation was successful, we can now
+		// modify the in memory representations. Set the outgoing
+		// circuit key on our pending circuit, add the same circuit to
+		// set of opened circuits, and add this circuit to the hash
+		// index.
+		circuit.Outgoing = &CircuitKey{}
+		*circuit.Outgoing = ks.OutKey
```

Comment:
what's this magic? 😮

Reply:
Initializes a new `CircuitKey` on the heap, the copies the contents of `ks.OutKey`, avoids maintaining references to keystones allocated in other part of the code base :)

Reply:
Interesting! Again, what do you think the performance gain is here?

Reply:
This isn't for performance, this is just a precaution to limit the number ways the outgoing key can be modified outside of the circuit map.

Reply:
For example, if we just do `circuit.Outgoing = &ks.OutKey`, the value `circuit.Outgoing` points to can change if the entry in the `keystones` array changes.  The link tries to reuse the same backing slice each batch, so that could cause problems if we were referencing into that slice